aten/src/ATen/StorageImpl.cpp

```diff
@@ -28,11 +28,12 @@ StorageImpl::StorageImpl(
           allocator,
           resizable) {}
 
-Type& StorageImpl::type() {
-  if (data_ptr.device().is_cuda()) {
-    return globalContext().getType(Backend::CUDA, scalar_type);
+namespace detail {
+Backend get_backend(StorageImpl* storage_impl) {
```

Comment:
I don't think this really addresses the issue.  The point is that there is not a one-to-one function from storage to backend -- don't write a function like this.  Why doesn't the usage in `checked_cast_storage` just take a DeviceType or something?  And the usage in getPyTypeObject you can just call get_backend in that same file.